opcua.cpp

```diff
@@ -719,6 +770,95 @@ void OPCUA::clearSubscription()
+{
+	// if filtering is disabled, always subscribe to given node
+	if(!getFilterEnabled())
+	{
+		Logger::getLogger()->info("%s:%d: FILTERING IS DISABLED", __FUNCTION__, __LINE__);
+		return true;
+	}
+
+	OPCUA::NodeFilterScope filterScope = getFilterScope();
+
+	// No check is required in case of a directly subscribed variable when filterScope is SCOPE_OBJECT
+	if( (nodeClass == OpcUa_NodeClass_Variable) && isDirectlySubscribed && 
+				 (filterScope == OPCUA::NodeFilterScope::SCOPE_OBJECT) )
+	{
+		Logger::getLogger()->info("Bypassing filtering check in case of directly subscribed variables when filterScope is SCOPE_OBJECT");
+		return true;
+	}
+
+	// No check is required in case of objects when filterScope is SCOPE_VARIABLE; just browse all objects
+	if( (nodeClass == OpcUa_NodeClass_Object) && (filterScope == OPCUA::NodeFilterScope::SCOPE_VARIABLE) )
+	{
+		Logger::getLogger()->info("Bypassing filtering check in case of objects when filterScope is SCOPE_VARIABLE");
+		return true;
+	}
+
+	bool includeNode = (getFilterAction() == OPCUA::NodeFilterAction::INCLUDE_NODES);
+	Logger::getLogger()->debug("getFilterAction()=%d, OPCUA::NodeFilterAction::INCLUDE_NODES=%d, includeNode=%s",
+									getFilterAction(), OPCUA::NodeFilterAction::INCLUDE_NODES, includeNode?"TRUE":"FALSE");
+	
+	bool scopeMatch = ((nodeClass == OpcUa_NodeClass_Object && 
+						(filterScope == OPCUA::NodeFilterScope::SCOPE_OBJECT || filterScope == OPCUA::NodeFilterScope::SCOPE_OBJECT_VARIABLE)) ||
+						(nodeClass == OpcUa_NodeClass_Variable));  // no need to check scope in case of variables; won't reject a variable irrespective of filter scope
+	
+	if (scopeMatch)
+	{
+		string filterRegex = getFilterRegex();
+		regex re(filterRegex);
+		bool match = std::regex_match(browseName, re);
+		Logger::getLogger()->info("filterAction=%s, filterScope=%s, nodeClass=%s, scopeMatch=%s, browseName=%s, filterRegex=%s, match=%s, subscribe=%s",
+										getFilterActionStr().c_str(), getFilterScopeStr().c_str(), nodeClassStr(nodeClass).c_str(), 
+										scopeMatch?"TRUE":"FALSE", browseName.c_str(), filterRegex.c_str(),
+										match?"TRUE":"FALSE", (includeNode == match)?"TRUE":"FALSE");
+		return (includeNode == match);
+		
+		// include - regex match - subscribe
+		// include - regex not matched - no subscribe
+
+		// exclude - regex match - no subscribe
+		// exclude - regex not matched - subscribe
+	}
+	else
+	{
+		Logger::getLogger()->info("filterAction=%s, filterScope=%s, nodeClass=%s, scopeMatch=%s, browseName=%s, subscribe=%s",
+										getFilterActionStr().c_str(), getFilterScopeStr().c_str(), nodeClassStr(nodeClass).c_str(), 
+										scopeMatch?"TRUE":"FALSE", browseName.c_str(), (!includeNode)?"TRUE":"FALSE");
+		return !includeNode;
+		// scope mismatch - include - can't include
+		// scope mismatch - exclude - ok to include
+	}
+}
+
+bool OPCUA::checkNode(std::string nodeStr, bool isDirectlySubscribed)
```

Comment:
Add a documentation header with parameters and return value.

Reply:
Added